src/Domain/Structure/ZCurve.cpp

```diff
@@ -0,0 +1,107 @@
+  // for the bit manipulation of the element index, we need to interleave the
+  // indices in each dimension in order according to how many bits are in the
+  // index representation. This variable stores the refinement level and
+  // dimension index in ascending order of refinement level, representing a
+  // permutation of the dimensions
+  // pair<refinement level, dim index> in order of ascending refinement
+  std::array<std::pair<size_t, size_t>, Dim>
+      dimension_by_highest_refinement_level;
+  for (size_t i = 0; i < Dim; ++i) {
+    dimension_by_highest_refinement_level.at(i) =
+        std::make_pair(element_id.segment_id(i).refinement_level(), i);
+  }
+  alg::sort(dimension_by_highest_refinement_level);
+
+  size_t element_order_index = 0;
+
+  // 'gap' the lowest refinement direction bits as:
+  // ... x1 x0 -> ... x1 0 0 x0,
+  // then bitwise or in 'gap'ed and shifted next-lowest refinement direction
+  // bits as:
+  // ... y2 y1 y0 -> ... y2 0 y1 x1 0 y0 x0
+  // then bitwise or in 'gap'ed and shifted highest-refinement direction bits
+  // as:
+  // ... z3 z2 z1 z0 -> z3 z2 y2 z1 y1 x1 z0 y0 x0
+  // note that we must skip refinement-level 0 dimensions as though they are
+  // not present
+  size_t leading_gap = 0;
+  for (size_t i = 0; i < Dim; ++i) {
+    const size_t id_to_gap_and_shift =
+        element_id
+            .segment_id(
+                gsl::at(dimension_by_highest_refinement_level, i).second)
+            .index();
+    size_t total_gap = leading_gap;
+    if (gsl::at(dimension_by_highest_refinement_level, i).first > 0) {
+      ++leading_gap;
+    }
+    for (size_t bit_index = 0;
+         bit_index < gsl::at(dimension_by_highest_refinement_level, i).first;
+         ++bit_index) {
+      // This operation will not overflow for our present use of `ElementId`s.
+      // This technique densely assigns an ElementID a unique size_t identifier
+      // determining the Morton curve order, and `ElementId` supports refinement
+      // levels such that a global index within a block will fit in a 64-bit
+      // unsigned integer.
+      element_order_index |=
+          ((id_to_gap_and_shift & two_to_the(bit_index)) << total_gap);
+      for (size_t j = 0; j < Dim; ++j) {
+        if (i != j and
+            bit_index + 1 <
+                gsl::at(dimension_by_highest_refinement_level, j).first) {
+          ++total_gap;
+        }
+      }
+    }
+  }
+  return element_order_index;
+}
+
+template <size_t Dim>
+std::vector<ElementId<Dim>> initial_element_ids_in_z_curve_order(
+    const size_t block_id, const std::array<size_t, Dim> initial_ref_levs,
+    const size_t grid_index) {
+  std::vector<ElementId<Dim>> element_ids =
+      initial_element_ids(block_id, initial_ref_levs, grid_index);
+  alg::sort(element_ids,
+            [](const ElementId<Dim>& lhs, const ElementId<Dim>& rhs) {
+              return z_curve_index(lhs) < z_curve_index(rhs);
+            });
+
+  return element_ids;
```

Comment:
Just delete this function and sort in-place where you need to. Shallow wrappers like this can actually make the code harder to understand because you think they do something fancy but they really don't :)